### ISSUE TITLE
Revisiting mistyped Assists

### ISSUE THREAD

**Issue by TheMathNinja** [mrcaseb/nflfastR#85]
Sorry if this should not be re-opened as a new issue. I wasn't sure if there was any notification given when I further commented on the previous closed issue. Feel free to close this if this is just clutter. But I believe this is a true issue, and not just a feature request. I just ran this code for 2020 Week 1:

pbp2020 <- readRDS(url('https://raw.githubusercontent.com/guga31bb/nflfastR-data/master/data/play_by_play_2020.rds'))
concern_players <- pbp2020 %>%
filter((solo_tackle_1_player_name == "K.Alexander" |
solo_tackle_2_player_name == "K.Alexander" |
assist_tackle_1_player_name == "K.Alexander" |
assist_tackle_2_player_name == "K.Alexander" |
assist_tackle_3_player_name == "K.Alexander" |
assist_tackle_4_player_name == "K.Alexander" |
solo_tackle_1_player_name == "J.Bosa" |
solo_tackle_2_player_name == "J.Bosa" |
assist_tackle_1_player_name == "J.Bosa" |
assist_tackle_2_player_name == "J.Bosa" |
assist_tackle_3_player_name == "J.Bosa" |
assist_tackle_4_player_name == "J.Bosa" |
solo_tackle_1_player_name == "B.Wagner" |
solo_tackle_2_player_name == "B.Wagner" |
assist_tackle_1_player_name == "B.Wagner" |
assist_tackle_2_player_name == "B.Wagner" |
assist_tackle_3_player_name == "B.Wagner" |
assist_tackle_4_player_name == "B.Wagner" |
solo_tackle_1_player_name == "L.David" |
solo_tackle_2_player_name == "L.David" |
assist_tackle_1_player_name == "L.David" |
assist_tackle_2_player_name == "L.David" |
assist_tackle_3_player_name == "L.David" |
assist_tackle_4_player_name == "L.David")) %>%
select(solo_tackle_1_player_name,
solo_tackle_2_player_name,
assist_tackle_1_player_name,
assist_tackle_2_player_name,
assist_tackle_3_player_name,
assist_tackle_4_player_name)
view(concern_players)

This gives the following outputs:
Kwon Alexander: 6 solos, 4 assists
Joey Bosa: 4 solos, 1 assist
Bobby Wagner: 1 solo, 6 assists
Lavonte David: 6 solos, 5 assists

However, Pro Football Reference (https://www.pro-football-reference.com/years/2020/defense.htm#defense::tackles_solo) and ESPN (https://www.espn.com/nfl/stats/player/_/view/defense/table/defensive/sort/soloTackles/dir/desc) and CBS Sports (https://www.cbssports.com/nfl/stats/player/defense/nfl/regular/qualifiers/?sortcol=tackles_cmb&sortdir=descending) give:
Kwon Alexander: 8 solos, 2 assists
Joey Bosa: 5 solos, 0 assists
Bobby Wagner: 2 solos, 5 assists
Lavonte David: 7 solos, 4 assists

where NFL.com gives:
Joey Bosa: 4 solos, 0 assists, 5 total
Bobby Wagner: 1 solo, 5 assists, 7 total
Lavonte David: 6 solos, 4 assists, 11 total

In all 4 player cases, nflfastR is over-counting Assists because it's counting stat_id 80 as an Assist, but it needs to be either:
1. Counting it as a solo (as PFR, ESPN, and CBS do), or
2. Counting it as its own separate category (as NFL.com does)
in order to be consistent with language from other sites.

In short, nobody classifies stat_id 80 an Assist, so I think this should be re-coded ASAP. I personally recommend making it its own category called "tackle_with_assist", but I think coding it as solo would also be a significant step forward. I prefer coding it as its own category so that nflfastR users can combine these three distinct categories to their liking for their own custom solo/assist categories which fit their format.

**Comment by TheMathNinja** [comment]
I should add one point of clarity: I think if stat_id 80 was re-coded from Assist to Solo for now, this would move from "Bug" to "Feature Request". The "tackle_with_assist" option is a bit more nuanced of an idea. It's only a bug currently because nobody calls stat_id 80 an assist.

**Comment by mrcaseb** [comment]
I have investigated this issue a little bit. It's pretty clear to me that assists indeed are counted incorrect compared to the NFL's stats sites. They are either same or higher depending on if the player recorded stat id 80.

See code output of the following chunck compared to the nfl.com screenshot

``` r
library(tidyverse)

pbp <- readRDS(url("https://raw.githubusercontent.com/guga31bb/nflfastR-data/master/data/play_by_play_2020.rds"))


solos <- pbp %>% 
  count(solo_tackle_1_player_name, solo_tackle_1_team) %>%
  arrange(desc(n)) %>% 
  filter(!is.na(solo_tackle_1_player_name)) %>% 
  rename(solos = n) %>% 
  slice_head(n = 6)

pbp %>% 
  filter(assist_tackle_1_player_name %in% solos$solo_tackle_1_player_name) %>% 
  count(assist_tackle_1_player_name, assist_tackle_1_team) %>% 
  rename(assist_tackle_1_n = n) %>% 
  left_join(
    pbp %>% 
      filter(assist_tackle_2_player_name %in% solos$solo_tackle_1_player_name) %>% 
      count(assist_tackle_2_player_name, assist_tackle_2_team) %>% 
      rename(assist_tackle_2_n = n),
    by = c("assist_tackle_1_player_name" = "assist_tackle_2_player_name", "assist_tackle_1_team" = "assist_tackle_2_team")
  ) %>% 
  mutate(
    assists = assist_tackle_1_n + assist_tackle_2_n
  ) %>% 
  select(-c(assist_tackle_1_n, assist_tackle_2_n)) %>% 
  right_join(
    solos, by = c("assist_tackle_1_player_name" = "solo_tackle_1_player_name", "assist_tackle_1_team" = "solo_tackle_1_team")
  ) %>% 
  arrange(desc(solos))
#> # A tibble: 6 x 4
#>   assist_tackle_1_player_name assist_tackle_1_team assists solos
#>   <chr>                       <chr>                  <int> <int>
#> 1 M.Kiser                     LA                         6    17
#> 2 C.Kirksey                   GB                         9    15
#> 3 E.Kendricks                 MIN                        8    15
#> 4 D.White                     TB                        12    14
#> 5 J.Adams                     SEA                        8    14
#> 6 C.Hayward Jr.               LAC                       NA    14
```

<sup>Created on 2020-09-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

![grafik](https://user-images.githubusercontent.com/38586519/93770048-f7353f80-fc1b-11ea-98de-5c4637faf982.png)

source: https://www.nfl.com/stats/player-stats/category/tackles/2020/REG/all/defensivesolotackles/DESC

The question is what to do with stat id 80. **I will not count it as solo tackle as those numbers appear to be correct.**

Here is the definition
![grafik](https://user-images.githubusercontent.com/38586519/93770558-a114cc00-fc1c-11ea-9afd-616704d10c83.png)

The only correct way is using a new variable called "tackle_with_assist" for stat id 80. But since we are in a running season I am rather against changing the output with additional variables.

My suggestion:
We change case 80 in `sum_play_stats()` to do nothing in that case instead of counting assisted tackles and add the "tackle_with_assist" variables in the next offseason.
What are your thoughts on that suggestion @guga31bb?

**Comment by guga31bb** [comment]
> We change case 80 in sum_play_stats() to do nothing in that case instead of counting assisted tackles and add the "tackle_with_assist" variables in the next offseason.

Yep this makes sense. I agree that we should not add additional columns until the season is over.

**Comment by TheMathNinja** [comment]
Is there any way for me to study combined tackles (meaning solos + stat id 80) with this current setup. Do I just have to use fast_scraper instead of the data repo? Currently in the data repo all tackles of stat id 80 are hidden unless they are registered as a TFL as far as I can tell. Any way I can see/analyze all stat id 80 that have been recorded this year if there's no variable for it yet?

**Comment by mrcaseb** [comment]
This code will return all plays with stat ID 80. It runs like 10 seconds or so but will take longer the more games are completed.

```r
`%>%` <- magrittr::`%>%`
get_stats <- function(id) {
  season <- substr(id, 1, 4)

  path <- "https://raw.githubusercontent.com/guga31bb/nflfastR-raw/master/raw"
  raw_data <- readRDS(url(glue::glue("{path}/{season}/{id}.rds")))

  plays <- raw_data$data$viewer$gameDetail$plays

  tidyr::unnest(plays %>% dplyr::select(-"yards"), cols = c("playStats")) %>%
    dplyr::mutate(
      game_id = id,
      yards = as.integer(.data$yards),
      statId = as.numeric(.data$statId),
      team.abbreviation = as.character(.data$team.abbreviation)
    ) %>%
    janitor::clean_names() %>%
    dplyr::select(
      game_id,
      play_id = order_sequence,
      desc = play_description_with_jersey_numbers,
      stat_id,
      yards,
      player_name,
      team_abbreviation,
      gsis_player_id
    ) %>%
    dplyr::filter(stat_id == 80)
}

ids <- nflfastR::fast_scraper_schedules(2020) %>% 
  dplyr::filter(!is.na(home_result)) %>% 
  dplyr::pull(game_id)

suppressWarnings(future::plan("multiprocess")) 

eighties <- furrr::future_map_dfr(ids, get_stats)

eighties <- eighties %>% 
  nflfastR::decode_player_ids()
```

```
> eighties %>% head(10)
# A tibble: 10 x 8
   game_id    play_id desc                              stat_id yards player_name team_abbreviati… gsis_player_id
   <chr>        <dbl> <chr>                               <dbl> <int> <chr>       <chr>            <chr>         
 1 2020_01_H…     693 (3:38) (Shotgun) 31-D.Williams l…      80     0 Z.Cunningh… HOU              00-0033547    
 2 2020_01_S…     777 (2:54) 73-M.Gono reported in as …      80     0 B.Wagner    SEA              00-0029255    
 3 2020_01_S…    4102 (6:15) 32-C.Carson right guard t…      80     0 D.Jones     ATL              00-0032388    
 4 2020_01_C…     744 (2:24) (Shotgun) 27-K.Hunt right…      80     0 C.Clark     BAL              00-0033294    
 5 2020_01_C…    1790 (:59) (Shotgun) 6-B.Mayfield pas…      80     0 D.Wolfe     BAL              00-0029586    
 6 2020_01_C…    3136 (1:44) (Shotgun) 8-L.Jackson pas…      80     0 M.Smith     CLE              00-0028180    
 7 2020_01_C…    3801 (7:14) 27-K.Hunt right tackle to…      80     0 C.Clark     BAL              00-0033294    
 8 2020_01_N…      94 (14:47) 26-L.Bell up the middle …      80     0 T.Murphy    BUF              00-0031300    
 9 2020_01_N…     137 (13:58) 7-B.Mann punts 58 yards …      80     0 R.Griffin   NYJ              00-0030108    
10 2020_01_N…     399 (9:37) 7-B.Mann punts 45 yards t…      80     0 A.Davis     NYJ              00-0036329  
```

**Comment by mrcaseb** [comment]
You are using nflfastR < v3

Best way is to run
```r
if (!requireNamespace("remotes", quietly = TRUE)) {install.packages("remotes")}
remotes::install_github("mrcaseb/nflfastR")
```
